### ISSUE TITLE
PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre() fails

### ISSUE THREAD

**Issue by sebastianbergmann** [sebastianbergmann/phpunit#2826]
```
PHPUnit 6.5-g1ce64f90a by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.10 with Xdebug 2.5.5
Configuration: /usr/local/src/phpunit/phpunit.xml

.............................................................   61 / 1559 (  3%)
.............................................................  122 / 1559 (  7%)
.............................................................  183 / 1559 ( 11%)
.............................................................  244 / 1559 ( 15%)
.............................................................  305 / 1559 ( 19%)
.............................................................  366 / 1559 ( 23%)
.............................................................  427 / 1559 ( 27%)
.............................................................  488 / 1559 ( 31%)
.............................................................  549 / 1559 ( 35%)
.............................................................  610 / 1559 ( 39%)
.............................................................  671 / 1559 ( 43%)
.............................................................  732 / 1559 ( 46%)
.............................................................  793 / 1559 ( 50%)
.............................................................  854 / 1559 ( 54%)
..........................................................ES.  915 / 1559 ( 58%)
.............................................................  976 / 1559 ( 62%)
............................................................. 1037 / 1559 ( 66%)
............................................................. 1098 / 1559 ( 70%)
............................................................. 1159 / 1559 ( 74%)
............................................................. 1220 / 1559 ( 78%)
............................................................. 1281 / 1559 ( 82%)
............................................................. 1342 / 1559 ( 86%)
............................................................. 1403 / 1559 ( 89%)
............................................................. 1464 / 1559 ( 93%)
............................................................. 1525 / 1559 ( 97%)
............S.....................                            1559 / 1559 (100%)

Time: 15.85 seconds, Memory: 14.00MB

There was 1 error:

1) PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre
Exception: Serialization of 'Closure' is not allowed

/usr/local/src/phpunit/vendor/sebastian/global-state/src/Snapshot.php:304
/usr/local/src/phpunit/vendor/sebastian/global-state/src/Snapshot.php:116
/usr/local/src/phpunit/src/Framework/TestCase.php:2312
/usr/local/src/phpunit/src/Framework/TestCase.php:2233
/usr/local/src/phpunit/src/Framework/TestCase.php:913
/usr/local/src/phpunit/src/Framework/TestResult.php:698
/usr/local/src/phpunit/src/Framework/TestCase.php:894
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

--

There were 2 skipped tests:

1) PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPost
This test depends on "PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre" to pass.

/usr/local/src/phpunit/src/Framework/TestCase.php:2056
/usr/local/src/phpunit/src/Framework/TestCase.php:781
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

2) /usr/local/src/phpunit/tests/Regression/GitHub/2591-separate-function-no-preserve-no-bootstrap.phpt
xdebug loaded

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

ERRORS!
Tests: 1559, Assertions: 2833, Errors: 1, Skipped: 2.
```

**Comment by sebastianbergmann** [comment]
Does not fail when run by itself:
```
./phpunit --filter testStaticAttributesBackupPre
PHPUnit 6.5-g1ce64f90a by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.10 with Xdebug 2.5.5
Configuration: /usr/local/src/phpunit/phpunit.xml

.                                                                   1 / 1 (100%)

Time: 204 ms, Memory: 10.00MB

OK (1 test, 0 assertions)
```

**Comment by sebastianbergmann** [comment]
@localheinz The `PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre()` test only fails when the tests that were added in 1ce64f90a27a563c9eeef8cf384aabfeddaf572a are executed before it:

```
$ git rev-parse --verify HEAD
1ce64f90a27a563c9eeef8cf384aabfeddaf572a
```

```
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean
```

```
./phpunit                  
PHPUnit 6.5-g1ce64f90a by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.10
Configuration: /usr/local/src/phpunit/phpunit.xml

.............................................................   61 / 1559 (  3%)
.............................................................  122 / 1559 (  7%)
.............................................................  183 / 1559 ( 11%)
.............................................................  244 / 1559 ( 15%)
.............................................................  305 / 1559 ( 19%)
.............................................................  366 / 1559 ( 23%)
.............................................................  427 / 1559 ( 27%)
.............................................................  488 / 1559 ( 31%)
.............................................................  549 / 1559 ( 35%)
.............................................................  610 / 1559 ( 39%)
.............................................................  671 / 1559 ( 43%)
.......................S.....................................  732 / 1559 ( 46%)
.............................................................  793 / 1559 ( 50%)
.............................................................  854 / 1559 ( 54%)
..........................................................ES.  915 / 1559 ( 58%)
.............................................................  976 / 1559 ( 62%)
............................................................. 1037 / 1559 ( 66%)
............................................................. 1098 / 1559 ( 70%)
............................................................. 1159 / 1559 ( 74%)
............................................................. 1220 / 1559 ( 78%)
............................................................. 1281 / 1559 ( 82%)
............................................................. 1342 / 1559 ( 86%)
............................................................. 1403 / 1559 ( 89%)
.............SS....................S......................... 1464 / 1559 ( 93%)
............................................................. 1525 / 1559 ( 97%)
...........S......................                            1559 / 1559 (100%)

Time: 10.56 seconds, Memory: 12.00MB

There was 1 error:

1) PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre
Exception: Serialization of 'Closure' is not allowed

/usr/local/src/phpunit/vendor/sebastian/global-state/src/Snapshot.php:304
/usr/local/src/phpunit/vendor/sebastian/global-state/src/Snapshot.php:116
/usr/local/src/phpunit/src/Framework/TestCase.php:2312
/usr/local/src/phpunit/src/Framework/TestCase.php:2233
/usr/local/src/phpunit/src/Framework/TestCase.php:913
/usr/local/src/phpunit/src/Framework/TestResult.php:698
/usr/local/src/phpunit/src/Framework/TestCase.php:894
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

--

There were 6 skipped tests:

1) PHPUnit\Framework\Constraint\ExceptionMessageRegExpTest::testMessageXdebugScreamCompatibility
Extension xdebug is required.

/usr/local/src/phpunit/src/Framework/Assert.php:2900
/usr/local/src/phpunit/src/Framework/TestCase.php:711
/usr/local/src/phpunit/src/Framework/TestCase.php:922
/usr/local/src/phpunit/src/Framework/TestResult.php:698
/usr/local/src/phpunit/src/Framework/TestCase.php:894

2) PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPost
This test depends on "PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre" to pass.

/usr/local/src/phpunit/src/Framework/TestCase.php:2056
/usr/local/src/phpunit/src/Framework/TestCase.php:781
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

3) /usr/local/src/phpunit/tests/TextUI/code-coverage-ignore.phpt
Extension xdebug is required.

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

4) /usr/local/src/phpunit/tests/TextUI/code-coverage-phpt.phpt
Extension xdebug is required.

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

5) /usr/local/src/phpunit/tests/TextUI/disable-code-coverage-ignore.phpt
Extension xdebug is required.

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

6) /usr/local/src/phpunit/tests/Regression/GitHub/2591-separate-function-no-preserve-no-bootstrap-xdebug.phpt
xdebug not loaded

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

ERRORS!
Tests: 1559, Assertions: 2831, Errors: 1, Skipped: 6.
```

```
$ rm tests/Framework/Constraint/Logical*Test.php
```

```
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	deleted:    tests/Framework/Constraint/LogicalAndTest.php
	deleted:    tests/Framework/Constraint/LogicalOrTest.php
	deleted:    tests/Framework/Constraint/LogicalXorTest.php

no changes added to commit (use "git add" and/or "git commit -a")
```

```
$ ./phpunit                                     
PHPUnit 6.5-g1ce64f90a by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.10
Configuration: /usr/local/src/phpunit/phpunit.xml

.............................................................   61 / 1556 (  3%)
.............................................................  122 / 1556 (  7%)
.............................................................  183 / 1556 ( 11%)
.............................................................  244 / 1556 ( 15%)
.............................................................  305 / 1556 ( 19%)
.............................................................  366 / 1556 ( 23%)
.............................................................  427 / 1556 ( 27%)
.............................................................  488 / 1556 ( 31%)
.............................................................  549 / 1556 ( 35%)
.............................................................  610 / 1556 ( 39%)
.............................................................  671 / 1556 ( 43%)
.......................S.....................................  732 / 1556 ( 47%)
.............................................................  793 / 1556 ( 50%)
.............................................................  854 / 1556 ( 54%)
.............................................................  915 / 1556 ( 58%)
.............................................................  976 / 1556 ( 62%)
............................................................. 1037 / 1556 ( 66%)
............................................................. 1098 / 1556 ( 70%)
............................................................. 1159 / 1556 ( 74%)
............................................................. 1220 / 1556 ( 78%)
............................................................. 1281 / 1556 ( 82%)
............................................................. 1342 / 1556 ( 86%)
............................................................. 1403 / 1556 ( 90%)
..........SS....................S............................ 1464 / 1556 ( 94%)
............................................................. 1525 / 1556 ( 98%)
........S......................                               1556 / 1556 (100%)

Time: 10.5 seconds, Memory: 12.00MB

There were 5 skipped tests:

1) PHPUnit\Framework\Constraint\ExceptionMessageRegExpTest::testMessageXdebugScreamCompatibility
Extension xdebug is required.

/usr/local/src/phpunit/src/Framework/Assert.php:2900
/usr/local/src/phpunit/src/Framework/TestCase.php:711
/usr/local/src/phpunit/src/Framework/TestCase.php:922
/usr/local/src/phpunit/src/Framework/TestResult.php:698
/usr/local/src/phpunit/src/Framework/TestCase.php:894

2) /usr/local/src/phpunit/tests/TextUI/code-coverage-ignore.phpt
Extension xdebug is required.

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

3) /usr/local/src/phpunit/tests/TextUI/code-coverage-phpt.phpt
Extension xdebug is required.

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

4) /usr/local/src/phpunit/tests/TextUI/disable-code-coverage-ignore.phpt
Extension xdebug is required.

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

5) /usr/local/src/phpunit/tests/Regression/GitHub/2591-separate-function-no-preserve-no-bootstrap-xdebug.phpt
xdebug not loaded

/usr/local/src/phpunit/src/Runner/PhptTestCase.php:252
/usr/local/src/phpunit/src/Runner/PhptTestCase.php:176
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/Framework/TestSuite.php:744
/usr/local/src/phpunit/src/TextUI/TestRunner.php:539
/usr/local/src/phpunit/src/TextUI/Command.php:195
/usr/local/src/phpunit/src/TextUI/Command.php:148

OK, but incomplete, skipped, or risky tests!
Tests: 1556, Assertions: 2812, Skipped: 5.
```
Looks like somehow, somewhere a `Closure` ends up in global state.

**Comment by localheinz** [comment]
@sebastianbergmann 

Taking a look!

**Comment by localheinz** [comment]
@sebastianbergmann 

Interestingly, when running the entire test suite, it passes only, when **all 3 tests** are removed or marked as skipped.

**Comment by localheinz** [comment]
@sebastianbergmann 

Modifying the tests to 

```php
public function testFromConstraintsReturnsConstraint()
{
    LogicalAnd::fromConstraints($this->createMock(Constraint::class));
    
    $this->assertNull(null);
}
```

results in the same error, then I realized that `Constraint` is an `abstract` class, thus the fix in #2829!

**Comment by sebastianbergmann** [comment]
I will have a look at #2829 in a second but I want to share this here for posterity's sake.

This
```
array(1) {
  ["Mock_Constraint_11824685"]=>
  object(Closure)#2874 (2) {
    ["static"]=>
    array(1) {
      ["reflectionClass"]=>
      object(ReflectionClass)#2854 (1) {
        ["name"]=>
        string(24) "Mock_Constraint_11824685"
      }
    }
    ["this"]=>
    object(Doctrine\Instantiator\Instantiator)#2607 (0) {
    }
  }
}
```
seems to be the source of `Closure` in question.

Under regular circumstances, when PHPUnit does not run its own test suite, this is not an issue as `Doctrine\Instantiator\Instantiator` is on the blacklist of `SebastianBergmann\GlobalState\Snapshot`.
